Report frequency

leanil opened the pull request:
Add frequency command to the bot
Any (tag,number) pair in the message is recognized as argument.

leanil committed:
Store and serve report frequency for projects

leanil committed:
Add frequency command to the bot

Any (tag,number) pair in the message is recognized as argument.

Resolves #32

paldys commented on client/src/components/leftPanel/LeftPanel.js:
```diff
@@ -16,6 +18,13 @@ class LeftPanel extends Component {
+      tag.isLate = isLate
+      parts[isLate ? 0 : 1].push(tag)
+    })
+    const partitionedTags = parts[0].concat(parts[1])
```
Uff.. :slightly_smiling_face: 

If you have used sort it would have been more readable. JavaScript's built-in sort is not stable in Edge. Let's add lodash to the project and use its sort which is stable. :slightly_smiling_face:

leanil replied:
I've decided to use partition, I think it's even more readable.

leanil committed:
Show late projects on the top with red name

A project is late if the last report came in more days ago,
than the project's frequency.
Archived projects can't be late.

leanil commented on bot/src/commands.js:
```diff
@@ -1,19 +1,32 @@
+      const tagFrequencyPattern = new RegExp(`${tagRegexPattern.source}\\s+(\\d+)`, 'g')
+      const tagsWithFrequency = []
+      let result
+      while ((result = tagFrequencyPattern.exec(message)) !== null) {
```
This would be nicer with node 12's [matchAll](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll), but we don't have node 12 on Now.

leanil commented on bot/src/db.js:
```diff
@@ -60,3 +60,17 @@ export async function archive(tags, status, ts) {
+
+export async function setFrequency(tagsWithFrequency, ts) {
+  const uniqueTags = {}
+  tagsWithFrequency.forEach(([tag, frequency]) => {uniqueTags[tag] = frequency})
```
The command can contain a tag multiple times, but the query can't touch the same row twice.
Could be done on server side, except aggregation function on frequency in group by would have to be min or max, which is kind of arbitrary.

paldys replied:
So.. Why don't we do this right away? :slightly_smiling_face: 

i.e. Instead of building an array for  `tagsWithFrequency`, you could make it an object, which is passed to `setFrequency`, whose keys are the tags and values are the frequencies.

paldys commented on client/src/components/leftPanel/LeftPanel.js:
```diff
@@ -16,6 +19,12 @@ class LeftPanel extends Component {
+    for (const tag of tags) {
+      tag.isLate =
+        moment().diff(parseTs(tag.lastTs), 'days') > tag.frequency &&
+        !tag.isArchived
```
(nit) Let's go with `!tag.isArchived && moment().diff(parseTs(tag.lastTs), 'days') > tag.frequency`

paldys commented:
One more thing which is more related to the previous task with archiving. Pleas
1. Remove `tags` parameter from command's `action` function. It makes sense only for `archive`.
2. This means you won't be needing the line `const tags = getTags..` in `handleCommands`.
3. Archive's action function should do `getTags` on the `message` parameter it got and handle it there.

leanil committed:
Add isCommand flag to the db and the bot

Mark tags on messages that were recognized as commands.
This makes it possible to filter out commands from different metrics
(count, time of last update), but still display them together with
the regular reports.
Projects with only commands (and no regular reports) are not shown.

leanil commented on bot/src/slack.js:
```diff
@@ -23,13 +23,12 @@ const addMessage = async (event) => {
 
   await upsertReport(report)
 
+  const isCommand = await handleCommands(event, web)
```
I went back to returning boolean, because now it doesn't change the program flow, and it makes the function of `handleCommands` cleaner. It just looks for executable commands, and returns whether it found any.

leanil commented on bot/src/commands.js:
```diff
@@ -50,5 +51,7 @@ export async function handleCommands(event, web) {
       timestamp: ts,
       name: 'heavy_check_mark',
     })
+    isCommand = true
```
This could also return early if it finds a command. But if I think if more commands match a message, it's more consistent to execute all of them, than only the first one.
